ziggurat-core-geoip/src/geoip.rs

```diff
@@ -0,0 +1,130 @@
+//use ipinfo::{IpInfo, IpInfoConfig};   // TODO(asmie): implement IpInfo provider
+use ip2location::{Record, DB};
+use ipgeolocate::{Locator, Service};
+
+/// List of supported geoip providers.
+pub enum Provider {
+    IpApiCom,
+    IpApiCo,
+    IpInfo,
+    Ip2Location,
+}
+
+/// Geoip provider service configuration.
+pub struct GeoIPService {
+    /// Geoip provider. Look at `Provider` enum for supported providers.
+    pub provider: Provider,
+    /// API key (paid or free) for online providers. For offline providers put here path to database file.
+    pub api_key: String,
+}
+
+/// IP information
+pub struct IPGeoInfo {
+    /// IP address
+    pub ip: String,
+    /// Country name (long name)
+    pub country: Option<String>,
+    /// City name
+    pub city: Option<String>,
+    /// IP address latitude
+    pub latitude: Option<f64>,
+    /// IP address longitude
+    pub longitude: Option<f64>,
+    /// Timezone of the IP
+    pub timezone: Option<String>,
+    /// ISP name (unavailable for some providers)
+    pub isp: Option<String>,
+}
+
+impl GeoIPService {
+    pub fn new(provider: Provider, api_key: String) -> Self {
+        Self { provider, api_key }
+    }
+
+    pub async fn lookup_ip(&self, ip: String) -> Result<IPGeoInfo, String> {
```

Comment:
from the below test:
`let ipgeo = geoip.lookup_ip("8.8.8.8".to_string()).await.unwrap();`
`let ipgeo = geoip.lookup("8.8.8.8".to_string()).await.unwrap();`  // suggestion

API doesn't need to mention `ip` twice unless we are planning to support additional lookup with something else (but then we should also reconsider if the crate name is okay - it seems IP oriented)
```suggestion
    pub async fn lookup(&self, ip: String) -> Result<IPGeoInfo, String> {
```